### PR TITLE
Updating JSON-RPC crates

### LINKED ISSUE #2448: Invalid Params error should be more descriptive.
I'm getting an invalid params error that is mysterious to me.  Could you help me debug it?

Here is a sample query:

``` bash
curl -H "Content-Type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"from":"0xfdea65c8e26263f6d9a1b5de9555d2931a33b825","data":"606060405260405160208061062e833981016040528080519060200190919050505b80600160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908302179055505b506105d28061005c6000396000f360606040526000357c01000000000000000000000000000000000000000000000000000000009004806326ea4bf614610068578063372c12b11461008557806355389c3e146100b8578063c2ed2b051461011c578063f851a440146101ec57610063565b610002565b3461000257610083600480803590602001909190505061022a565b005b34610002576100a0600480803590602001909190505061031d565b60405180821515815260200191505060405180910390f35b346100025761011a6004808035906020019091908035906020019082018035906020019191908080601f016020809104026020016040519081016040528093929190818152602001838380828437820191505050505050909091905050610342565b005b34610002576101376004808035906020019091905050610563565b604051808373ffffffffffffffffffffffffffffffffffffffff168152602001806020018281038252838181546001816001161561010002031660029004815260200191508054600181600116156101000203166002900480156101dc5780601f106101b1576101008083540402835291602001916101dc565b820191906000526020600020905b8154815290600101906020018083116101bf57829003601f168201915b5050935050505060405180910390f35b34610002576101fe60048050506105ac565b604051808273ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b338073ffffffffffffffffffffffffffffffffffffffff16600160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1614151561028757610002565b6001600060005060008473ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060006101000a81548160ff021916908302179055507f078cba54125a07a375c22702c366fcb4d1029bafd27b1d6e93d813b1d103b6cb82604051808273ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390a15b5b5050565b600060005060205280600052604060002060009150909054906101000a900460ff1681565b3360001515600060005060008373ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060009054906101000a900460ff161515141561038e57610002565b604060405190810160405280338152602001838152602001506002600050600085815260200190815260200160002060005060008201518160000160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908302179055506020820151816001016000509080519060200190828054600181600116156101000203166002900490600052602060002090601f016020900481019282601f1061044657805160ff1916838001178555610477565b82800160010185558215610477579182015b82811115610476578251826000505591602001919060010190610458565b5b5090506104a29190610484565b8082111561049e5760008181506000905550600101610484565b5090565b50509050507f2ebfece40746549a3b7549b22060ca8d895e54ac0096edfb3aaee7db4acdc0ee833384604051808481526020018373ffffffffffffffffffffffffffffffffffffffff168152602001806020018281038252838181518152602001915080519060200190808383829060006004602084601f0104600302600f01f150905090810190601f16801561054d5780820380516001836020036101000a031916815260200191505b5094505050505060405180910390a15b5b505050565b60026000506020528060005260406000206000915090508060000160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff169080600101600050905082565b600160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16815600000000000000000000000058f3d6e28fa4ad86cdf2223e342f54fe00000000","gas":"0x47b760"}],"id":1475548134452321}' http://localhost:8545
```

If you point this at your own Parity instance like mine, you'll get back a reply like this:

``` json
{"jsonrpc":"2.0","error":{"code":-32602,"message":"Invalid params","data":null},"id":1475549061808573}
```

I'm not sure what's wrong with this request, geth handles it fine.

### LINKED ISSUE #4476: Flag --jsonrpc-cors should be improved to pre-process URL
`--jsonrpc-cors "http://ethereum.github.io/"` will not work, instead you need to correctly specify `--jsonrpc-cors "http://ethereum.github.io"`

Also https://github.com/ethcore/parity/wiki/Private-development-chain suggests `--jsonrpc-cors "https://ethereum.github.io/browser-solidity"`, and AFAIK should not be `https` and will not work due `/browser-solidity` in URL.

### LINKED ISSUE #4908: Avoid building mio 5 times
Currently, we are using 5 different versions of `mio` (!!!!!). We could significantly reduce total build time if we use a single one instead

### PR DESCRIPTION
**Pull request by tomusdrw**
- Contains new IPC implementation for Unix and Windows (using tokio)
- Contains changes required for #4474  and #884
- Proper error messages for invalid ser/de
- `data: null` no longer serialized for RPC error messages

Closes #2448 
Closes #4908 
Closes #4476 

Possibly fixes: #4783

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -26,7 +26,7 @@ app_dirs = "1.1.1"
 fdlimit = "0.1"
 hyper = { default-features = false, git = "https://github.com/ethcore/hyper" }
 ctrlc = { git = "https://github.com/ethcore/rust-ctrlc.git" }
-jsonrpc-core = { git = "https://github.com/ethcore/jsonrpc.git", branch = "parity-1.6" }
```
Comment (NikVolf) [inline Cargo.toml]: I think we should use dedicated branch for referencing from parity, maybe `parity-1.7` now

File: Cargo.toml [hunk 2]
```diff
@@ -26,7 +26,7 @@ app_dirs = "1.1.1"
...
+jsonrpc-core = { git = "https://github.com/ethcore/jsonrpc.git", branch = "master" }
```
Comment (NikVolf) [inline Cargo.toml]: I think we should use dedicated branch for referencing from parity, maybe `parity-1.7` now

Comment (tomusdrw) [inline Cargo.toml]: Addressed.

**Comment by NikVolf** [comment]
also addresses this
https://github.com/paritytech/parity/issues/4750

File: ethcore/light/src/on_demand/mod.rs [hunk 3]
```diff
@@ -18,6 +18,9 @@
 //! The request service is implemented using Futures. Higher level request handlers
 //! will take the raw data received here and extract meaningful results from it.
 
+// TODO [ToDr] Suppressing deprecation warnings. Rob will fix the API anyway.
+#![allow(deprecated)]
```
Comment (rphmeier) [inline ethcore/light/src/on_demand/mod.rs]: indeed :)

**Pull request merged**